File: sdk_v2/cs/src/OpenAI/AudioClient.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Microsoft.AI.Foundry.Local;

using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Threading.Channels;

using Betalgo.Ranul.OpenAI.ObjectModels.RequestModels;
using Betalgo.Ranul.OpenAI.ObjectModels.ResponseModels;

using Microsoft.AI.Foundry.Local.Detail;
using Microsoft.AI.Foundry.Local.Detail.Interop;
using Microsoft.AI.Foundry.Local.Detail.Native;
using Microsoft.AI.Foundry.Local.OpenAI;
using Microsoft.Extensions.Logging;

using NativeModel = Microsoft.AI.Foundry.Local.Detail.Native.Model;
using NativeSession = Microsoft.AI.Foundry.Local.Detail.Native.Session;

/// <summary>
/// Audio Client that uses the OpenAI API.
/// Implemented using Betalgo.Ranul.OpenAI SDK types.
/// </summary>
public class OpenAIAudioClient
{
    private readonly string _modelId;
    private readonly NativeModel _nativeModel;

    private readonly ILogger _logger = FoundryLocalManager.Instance.Logger;

    internal OpenAIAudioClient(string modelId, NativeModel nativeModel)
    {
        _modelId = modelId;
        _nativeModel = nativeModel;
    }

    /// <summary>
    /// Settings that are supported by Foundry Local
    /// </summary>
    public record AudioSettings
    {
        public string? Language { get; set; }
        public float? Temperature { get; set; }
    }

    /// <summary>
    /// Settings to use for audio transcription using this client.
    /// </summary>
    public AudioSettings Settings { get; } = new();

    /// <summary>
    /// Transcribe audio from a file.
    /// </summary>
    /// <param name="audioFilePath">
    /// Path to file containing audio recording.
    /// Supported formats: mp3
    /// </param>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Transcription response.</returns>
    public async Task<AudioCreateTranscriptionResponse> TranscribeAudioAsync(string audioFilePath,
                                                                             CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(() => TranscribeAudioImplAsync(audioFilePath, ct),
                                                     "Error during audio transcription.", _logger)
                                                    .ConfigureAwait(false);
    }

    /// <summary>
    /// Transcribe audio from a file with streamed output.
    /// </summary>
    /// <param name="audioFilePath">
    /// Path to file containing audio recording.
    /// Supported formats: mp3
    /// </param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>An asynchronous enumerable of transcription responses.</returns>
    public async IAsyncEnumerable<AudioCreateTranscriptionResponse> TranscribeAudioStreamingAsync(
        string audioFilePath, [EnumeratorCancellation] CancellationToken ct)
    {
        var enumerable = Utils.CallWithExceptionHandling(
            () => TranscribeAudioStreamingImplAsync(audioFilePath, ct),
            "Error during streaming audio transcription.", _logger).ConfigureAwait(false);

        await foreach (var item in enumerable)
        {
            yield return item;
        }
    }

    /// <summary>
    /// Create a real-time streaming transcription session.
    /// Audio data is pushed in as PCM chunks and transcription results are returned as an async stream.
    /// </summary>
    /// <returns>A streaming session that must be disposed when done.</returns>
    public LiveAudioTranscriptionSession CreateLiveTranscriptionSession()
    {
        return new LiveAudioTranscriptionSession(_modelId, _nativeModel);
    }

    private async Task<AudioCreateTranscriptionResponse> TranscribeAudioImplAsync(string audioFilePath,
                                                                                  CancellationToken? ct)
    {
        var openaiRequest = AudioTranscriptionCreateRequestExtended.FromUserInput(_modelId, audioFilePath, Settings);
        var requestJson = openaiRequest.ToJson();

        return await Task.Run(() =>
        {
            using var session = new NativeSession(_nativeModel);
            using var jsonItem = new TextItem(requestJson, TextItemType.OpenAIJson);
            using var request = new Request();
            request.AddItem(jsonItem);

            var responsePtr = session.ProcessRequest(request.Ptr);
            using var response = new Response(responsePtr);

            using var responseItem = response.GetItem(0);
            var responseJson = ((TextItem)responseItem).Text;

            return JsonSerializer.Deserialize(responseJson,
                       JsonSerializationContext.Default.AudioCreateTranscriptionResponse)
                   ?? throw new FoundryLocalException("Failed to deserialize audio transcription response.");
        }).ConfigureAwait(false);
    }

    private async IAsyncEnumerable<AudioCreateTranscriptionResponse> TranscribeAudioStreamingImplAsync(
        string audioFilePath, [EnumeratorCancellation] CancellationToken ct)
    {
        var openaiRequest = AudioTranscriptionCreateRequestExtended.FromUserInput(_modelId, audioFilePath, Settings);
        var requestJson = openaiRequest.ToJson();

        var channel = Channel.CreateUnbounded<AudioCreateTranscriptionResponse>(
                        new UnboundedChannelOptions
                        {
                            SingleWriter = true,
                            SingleReader = true,
                            AllowSynchronousContinuations = true
                        });

        _ = Task.Run(() =>
        {
            try
            {
                using var session = new NativeSession(_nativeModel);

                FlStreamingCallback streamingCallback = (FlStreamingCallbackData data, IntPtr userData) =>
                {
                    try
                    {
                        if (data.ItemQueue != IntPtr.Zero)
                        {
                            while (Api.Item.QueueTryPop(data.ItemQueue, out var itemPtr))
                            {
                                using var item = Item.FromNative(itemPtr, ownsHandle: true);
                                var responseJson = ((TextItem)item).Text;

                                var chunk = JsonSerializer.Deserialize(responseJson,
                                    JsonSerializationContext.Default.AudioCreateTranscriptionResponse);

                                if (chunk != null)
                                {
                                    channel.Writer.TryWrite(chunk);
                                }
                            }
                        }
                    }
                    catch (Exception ex)
                    {
                        channel.Writer.TryComplete(
                            new FoundryLocalException("Error processing streaming audio transcription callback data.", ex, _logger));
                    }

                    return ct.IsCancellationRequested ? 1 : 0;
                };

                session.SetStreamingCallback(streamingCallback);

                using var jsonItem = new TextItem(requestJson, TextItemType.OpenAIJson);
                using var request = new Request();
                request.AddItem(jsonItem);

                var responsePtr = session.ProcessRequest(request.Ptr);
                Api.Inference.ResponseRelease(responsePtr);

                channel.Writer.TryComplete();
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                channel.Writer.TryComplete(
                    new FoundryLocalException("Error executing streaming audio transcription.", ex, _logger));
            }
            catch (OperationCanceledException)
            {
                channel.Writer.TryComplete();
            }
        }, ct);

        await foreach (var item in channel.Reader.ReadAllAsync(ct))
        {
            yield return item;
        }
    }
}

File: sdk_v2/python/src/openai/audio_client.py
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------

from __future__ import annotations

import json
import logging
from dataclasses import dataclass, field
from typing import Callable, Optional

from src.detail.core_interop import CoreInterop, InteropRequest

logger = logging.getLogger(__name__)


class AudioSettings:
    """Settings supported by Foundry Local for audio transcription.

    Attributes:
        language: Language of the audio (e.g. ``"en"``).
        temperature: Sampling temperature (0.0 for deterministic results).
    """

    def __init__(
        self,
        language: Optional[str] = None,
        temperature: Optional[float] = None,
    ):
        self.language = language
        self.temperature = temperature


@dataclass
class AudioTranscriptionResponse:
    """Response from an audio transcription request.

    Attributes:
        text: The transcribed text.
    """

    text: str


class AudioClient:
    """OpenAI-compatible audio transcription client backed by Foundry Local Core.

    Supports non-streaming and streaming transcription of audio files.

    Attributes:
        model_id: The ID of the loaded Whisper model variant.
        settings: Tunable ``AudioSettings`` (language, temperature).
    """

    def __init__(self, model_id: str, core_interop: CoreInterop):
        self.model_id = model_id
        self.settings = AudioSettings()
        self._core_interop = core_interop

    @staticmethod
    def _validate_audio_file_path(audio_file_path: str) -> None:
        """Validate that the audio file path is a non-empty string."""
        if not isinstance(audio_file_path, str) or audio_file_path.strip() == "":
            raise ValueError("Audio file path must be a non-empty string.")

    def _create_request_json(self, audio_file_path: str) -> str:
        """Build the JSON payload for the ``audio_transcribe`` native command."""
        request: dict = {
            "Model": self.model_id,
            "FileName": audio_file_path,
        }

        metadata: dict[str, str] = {}

        if self.settings.language is not None:
            request["Language"] = self.settings.language
            metadata["language"] = self.settings.language

        if self.settings.temperature is not None:
            request["Temperature"] = self.settings.temperature
            metadata["temperature"] = str(self.settings.temperature)

        if metadata:
            request["metadata"] = metadata

        return json.dumps(request)

    def transcribe(self, audio_file_path: str) -> AudioTranscriptionResponse:
        """Transcribe an audio file (non-streaming).

        Args:
            audio_file_path: Path to the audio file to transcribe.

        Returns:
            An ``AudioTranscriptionResponse`` containing the transcribed text.

        Raises:
            ValueError: If *audio_file_path* is empty or the native command fails.
        """
        self._validate_audio_file_path(audio_file_path)

        request_json = self._create_request_json(audio_file_path)
        request = InteropRequest(params={"OpenAICreateRequest": request_json})

        response = self._core_interop.execute_command("audio_transcribe", request)
        if response.error is not None:
            raise ValueError(
                f"Audio transcription failed for model '{self.model_id}': {response.error}"
            )

        data = json.loads(response.data)
        return AudioTranscriptionResponse(text=data.get("text", ""))

    def transcribe_streaming(
        self,
        audio_file_path: str,
        callback: Callable[[AudioTranscriptionResponse], None],
    ) -> None:
        """Transcribe an audio file with streaming chunks.

        Each chunk is passed to *callback* as an ``AudioTranscriptionResponse``.

        Args:
            audio_file_path: Path to the audio file to transcribe.
            callback: Called with each incremental transcription chunk.

        Raises:
            ValueError: If *audio_file_path* is empty or the native command fails.
            TypeError: If *callback* is not callable.
        """
        self._validate_audio_file_path(audio_file_path)

        if not callable(callback):
            raise TypeError("Callback must be a valid function.")

        request_json = self._create_request_json(audio_file_path)
        request = InteropRequest(params={"OpenAICreateRequest": request_json})

        def callback_handler(chunk_str: str):
            chunk_data = json.loads(chunk_str)
            chunk = AudioTranscriptionResponse(text=chunk_data.get("text", ""))
            callback(chunk)

        response = self._core_interop.execute_command_with_callback(
            "audio_transcribe", request, callback_handler
        )
        if response.error is not None:
            raise ValueError(
                f"Streaming audio transcription failed for model '{self.model_id}': {response.error}"
            )